### ISSUE TITLE
undefined method fetch for nil:NilClass (in disable_password_restrictions)

### ISSUE THREAD

**Issue by loenk2x** [Parallels/vagrant-parallels#219]
When i do vagrant up, this occurs:

```
$ vagrant up
Bringing machine 'default' up with 'parallels' provider...
/Users/tituswiguno/.vagrant.d/gems/gems/vagrant-parallels-1.5.0/lib/vagrant-parallels/driver/pd_10.rb:54:in `disable_password_restrictions': undefined method `fetch' for nil:NilClass (NoMethodError)
	from /Users/tituswiguno/.vagrant.d/gems/gems/vagrant-parallels-1.5.0/lib/vagrant-parallels/action/import.rb:20:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /Users/tituswiguno/.vagrant.d/gems/gems/vagrant-parallels-1.5.0/lib/vagrant-parallels/action/customize.rb:39:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:95:in `block in finalize_action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/runner.rb:66:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/runner.rb:66:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builtin/call.rb:53:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builtin/config_validate.rb:25:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:95:in `block in finalize_action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builtin/handle_box.rb:56:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:95:in `block in finalize_action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/runner.rb:66:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/runner.rb:66:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builtin/call.rb:53:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/runner.rb:66:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/runner.rb:66:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/machine.rb:214:in `action_raw'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/machine.rb:191:in `block in action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/environment.rb:516:in `lock'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/machine.rb:178:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/machine.rb:178:in `action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/batch_action.rb:82:in `block (2 levels) in run'
```

I'm using latest version of vagrant-parallels, vagrant, with Parallels Desktop 10.3.0 (29227) on iMac OS X 10.11.

This problem is not happening on my MacBook Pro with all softwares on the same version.

**Comment by legal90** [comment]
Is this issue reproduced constantly, for each `vagarnt up`?
If so, please show the output of this command: `prlsrvctl info --json`

**Comment by loenk2x** [comment]
Yes. It is.
Here you go.

```
{
	"ID": "941d0f8a-db4d-4055-a6d6-8530cf81f562",
	"Hostname": "127.0.0.1",
	"Version": "Desktop 10.3.0-29227",
	"OS": "Mac OS X 10.11(15A284)",
	"Started as service": "off",
	"VM home": "/Users/tituswiguno/Documents/Parallels",
	"Memory limit": {
		"mode": "auto"
	},
	"Minimal security level": "low",
	"Manage settings for new users": "allow",
	"CEP mechanism": "off",
	"Backup path": "/Users/Shared/Parallels/Backups",
	"Backup temporary directory": "",
	"Backup timeout": 3600,
	"Default encryption plugin": "<parallels-default-plugin>",
	"Verbose log": "off",
	"Allow mobile clients": "off",
	"Proxy connection state": "disconnected",
	"Direct connection": "off",
	"Log rotation": "on",
	"Advanced security mode": "off",
	"Proxy manager URL": "https://pax-manager.myparallels.com/xmlrpc/rpc.do",
	"Web portal domain": "parallels.com",
	"Host ID": "",
	"Allow attach screenshots": "on",
	"License": {
		"state": "valid",
		"key": "045-18186-64852-23826-20564-38465",
		"restricted": "false"
	},
	"Hardware Id": "",
	"Hardware info": {
		"/dev/disk3": {
			"name": "TOSHIBA External USB 3.0 (disk3)",
			"type": "hdd"
		},
		"/dev/disk3s1": {
			"name": "EFI System Partition",
			"type": "hdd-part"
		},
		"/dev/disk3s2": {
			"name": "Untitled",
			"type": "hdd-part"
		},
		"/dev/disk4": {
			"name": "WD My Book 1140 (disk4)",
			"type": "hdd"
		},
		"/dev/disk4s1": {
			"name": "EFI System Partition",
			"type": "hdd-part"
		},
		"/dev/disk4s2": {
			"name": "LoenkZ-2TB",
			"type": "hdd-part"
		},
		"/dev/disk4s3": {
			"name": "Booter",
			"type": "hdd-part"
		},
		"IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/SATA@1F,2/AppleIntelPchSeriesAHCI/PRT1@1/IOAHCIDevice@0/IOAHCISerialATAPI/IOSCSIPeripheralDeviceNub/IOSCSIPeripheralDeviceType05/IODVDServices": {
			"name": "OPTIARC DVD RW AD-5680H",
			"type": "cdrom"
		},
		"OService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/SATA@1F,2/AppleIntelPchSeriesAHCI/PRT1@1/IOAHCIDevice@0/IOAHCISerialATAPI/IOSCSIPeripheralDeviceNub/IOSCSIPeripheralDeviceType05/IODVDServices": {
			"name": "TSSTcorp CDDVDW SE-208DB",
			"type": "cdrom"
		},
		"en0": {
			"name": "Ethernet",
			"type": "net"
		},
		"en1": {
			"name": "Wi-Fi",
			"type": "net"
		},
		"p2p0": {
			"name": "p2p0",
			"type": "net"
		},
		"vnic0": {
			"name": "vnic0",
			"type": "net"
		},
		"vnic1": {
			"name": "vnic1",
			"type": "net"
		},
		"fd110000|05ac|8502|high|--|8TA7101MSDDQ4A00": {
			"name": "Apple Built-in iSight",
			"type": "usb"
		},
		"fa140000|1058|1140|high|--|574D43344D32333233373139": {
			"name": "My Book 1140",
			"type": "usb"
		},
		"fa120000|05ac|8403|high|--|000000009833": {
			"name": "Apple Internal Memory Card Reader",
			"type": "usb"
		},
		"fd120000|05ac|8242|low|--|Empty": {
			"name": "Apple IR Receiver",
			"type": "usb"
		},
		"fa111000|05ac|8215|full|--|78CA39BE8091": {
			"name": "Apple Bluetooth USB Host Controller",
			"type": "usb"
		},
		"fd142000|04e8|6031|high|--|30303030303030303030303030303030": {
			"name": "External USB 3.0",
			"type": "usb"
		},
		"fd141100|046d|c52b|full|KM|Empty": {
			"name": "USB Receiver",
			"type": "usb"
		},
		"fd144000|04f9|0253|full|--|BROA2F355167": {
			"name": "DCP-J125",
			"type": "usb"
		},
		"fd141200|046d|c21f|full|--|Empty": {
			"name": "Wireless Gamepad F710",
			"type": "usb"
		},
		"fd141300|0e8d|1806|high|--|R91J6YADB0042C ": {
			"name": "MT1806 ",
			"type": "usb"
		},
		"/dev/cu.BlackBerry9330-BlackBer": {
			"name": "/dev/cu.BlackBerry9330-BlackBer",
			"type": "serial"
		},
		"/dev/cu.BlackBerry9330-BlackBer-1": {
			"name": "/dev/cu.BlackBerry9330-BlackBer-1",
			"type": "serial"
		},
		"/dev/cu.LLZ-MobileHQ-Bluetooth-": {
			"name": "/dev/cu.LLZ-MobileHQ-Bluetooth-",
			"type": "serial"
		},
		"/dev/cu.Bluetooth-Incoming-Port": {
			"name": "/dev/cu.Bluetooth-Incoming-Port",
			"type": "serial"
		}
	}
}
```

**Comment by legal90** [comment]
Hm... Your JSON looks absolutely valid and it should be parsed successfully. Seems like I have to look at the debug log.
Please, reproduce this issue again with getting a debug log:
```
$ vagrant up --debug 2>./debug.log
```

Put the debug log on Gist/Pastebin or send it to me via e-mail (specified in my profile).

**Comment by loenk2x** [comment]
Here's the log gist.
https://gist.github.com/fd4b599b0e87f16965a1.git

**Comment by legal90** [comment]
@loenk2x Sorry, it's unavailable (404). May be, this is a permission problem?

**Comment by loenk2x** [comment]
Oops, sorry.
Please try now.

**Comment by legal90** [comment]
It is so weird... I've even tried to parse json from your debug output and it works fine.
I'd like to ask you about one trick, which should help us:

1) Make sure you are using the latest version of vagrant-parallels plugin (`vagrant plugin list`)
2) Edit the file ~/.vagrant.d/gems/gems/vagrant-parallels-1.5.0/lib/vagrant-parallels/driver/base.rb
3) In the line 673 there should be this block:
```ruby
        def json(default=nil)
          data = yield
          begin
            JSON.parse(data)
          rescue JSON::ParserError
            # Try to cleanup the data and parse it again [GH-204]
            data = data[/(\{.*\}|\[.*\])/m]
            JSON.parse(data) rescue default
          end
        end
```
replace it temporary with this one:

```ruby
        def json(default=nil)
          data = yield
          JSON.parse(data)
        end
```
4) Reproduce the issue again - run `vagrant up` 
5) Show the displayed error message (or stack trace). I guess there will be some JSON-related exception.

Thanks

**Comment by loenk2x** [comment]
Yeah, it is weird.
Here's the log: https://gist.github.com/4f2989de33ace4569034.git

For what it's worth, this vagrant just started acting weird since today; yesterday everything was normal. Not sure what changes...

**Comment by legal90** [comment]
If it's possible, could you try to reinstall Vagrant? https://docs.vagrantup.com/v2/installation/uninstallation.html

It's really strange that embedded JSON module couldn't parse the valid string.

**Comment by loenk2x** [comment]
I did. Already 3 or 4 times, as far as i remember.

I reinstalled Vagrant, all plugins, and even Parallels Desktop. Restarted several times too. And reset the PRAM too.
...No luck...

I wonder what the heck is wrong with this.

**Comment by loenk2x** [comment]
Ok, this problem is resolved. I'm not sure why, but disconnecting my external DVD Drive did the work. And I'm really confused; that DVD drive was there since ages and it never went haywire.

Thanks for the help.

**Comment by retspen** [comment]
I had the same problem. That's little part my <code>prlsrvctl info --json</code>

```
               ...
                "en0": {
                        "name": "Ethernet^A",
                        "type": "net"
                },
                "en1": {
                        "name": "Wi-Fi",
                        "type": "net"
                },
                ...
```

<code>^A</code> - means empty symbol (maybe space). I removed empty symbol in Network Interface name and now it work fine.

**Comment by legal90** [comment]
It mostly like some control characters have been displayed in the output of `prlsrvctl`. 
In the @retspen's example it is actually so. `^A` (or `u0001` in Unicode) is not supported by JSON and it causes an exception while JSON parsing. 

I've prepared a workaround: #223. I hope, it will help in the most os cases like these.
Or, at least the readable error message will be displayed.

**Comment by legal90** [comment]
The workaround has been released in [v1.5.1](https://github.com/Parallels/vagrant-parallels/releases/tag/v1.5.1)
